### PR TITLE
WIP: Draft version constraint generation

### PR DESCRIPTION
**Pull request by Gertjan423**
First draft version of static verification tool.
- Reads DAR files.
- Partially evaluates the code.
- Generates constraints for all fields, as the tool does not yet read annotations regarding which fields should be tracked.

Main TODOs at the moment:
- Implement some undefined's and TODO's throughout the code.
- Take additional inputs or read annotations to determine which fields should be tracked.
- Extend with additional language constructs (recursion).
- Additional testing.

### CODE REVIEW

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    -- Perhaps we do need to define our own set of errors.
+    -- _                      -> throwError (EExpectedRecordType ty)
+
+genForExercise :: MonadDelta m => (Qualified TypeConName) -> ChoiceName
+               -> Expr -> Maybe Expr -> Expr
+               -> m GenOutput
+genForExercise tem ch cid par arg = do
+  cidOut <- genExpr cid
+  argOut <- genExpr arg
+  -- TODO: Take possibility into account that the choice is not found?
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:187]: That would be caught by the LF typechecker, so no need to do more than crash here.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    argExpr@(ERecCon _ fs) -> return (GenOutput (EUpdate (UCreate tem argExpr))
+                                      -- TODO: We could potentially filter here
+                                      -- to only store the interesting fields?
+                                      (set usCre [UpdCreate tem fs] emptyUpdateSet)
+                                      (argOut ^. goDel))
+    _                      -> throwError EEnumTypeWithParams
+    -- TODO: This is a random error, as we do not have access to the expected
+    -- type, which we need to constructed the error we really want.
+    -- Perhaps we do need to define our own set of errors.
+    -- _                      -> throwError (EExpectedRecordType ty)
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:179]: We generally format to be somewhat friendly to diffs. So no alignment, line breaks and indentation such that the amount of indentation does not depend on identifier length. Not a hard rule but I like being somewhat consistent here.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+genTemplate :: MonadDelta m => Template -> m Delta
+genTemplate = undefined -- TODO
+
+genExpr :: MonadDelta m => Expr -> m GenOutput
+genExpr = \case
+  ETmApp fun arg  -> genForTmApp fun arg
+  ETyApp expr typ -> genForTyApp expr typ
+  EVar name       -> genForVar name
+  EVal w          -> genForVal w
+  EUpdate (UCreate tem arg)              -> genForCreate tem arg
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:128]: This doesn’t seem quite right to me. You are conflating evaluation and execution afaict. `Update` works like `IO` in Haskell. Evaluating an `Update` expression does not do anything. The way this is implemented at runtime is by passing around an opaque token (similar to the RealWorld token in Haskell) so `Update` expressions are actually functions. You can probably do something similar for your analysis: If you start from a choice you look at the update expression evaluated to the token. If you just see an `Update` expression that is not applied to the token, you don’t look inside.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+data UpdateSet = UpdateSet
+  { _usCre :: ![UpdCreate]
+    -- ^ The list of create updates.
+  , _usArc :: ![UpdArchive]
+    -- ^ The list of archive updates.
+  , _usCho :: ![UpdChoice]
+    -- ^ The list of choice updates.
+  }
+
+makeLenses ''UpdateSet
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:62]: We generally avoid TemplateHaskell for the most part. Part of that is just a code style guideline since we generally try write fairly simple Haskell code but for TemplateHaskell there is also a technical resaon: We use a statically linked GHC (I won’t go into the reasons here but I’m happy to expand on that in another context) which makes TemplateHaskell very slow. So if you really do want to use template haskell. You are better off putting all the template Haskell code into a separate Bazel target that changes rarely (Bazel caches at the granularity of individual targets). Otherwise, you will suffer from long compile times while working on this.

However, we also mostly avoid lenses. This is not quite as strict and there is not really a technical reason for this but I’m not aware of any module that uses them as pervasively as you use them here.

So overall, I think you’re better off removing the TH lens generation, switch away from using lenses by default for everything and potentially define a few lenses by hand for the things where they do provide a significant benefit.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+emptyUpdateSet :: UpdateSet
+emptyUpdateSet = UpdateSet [] [] []
+
+concatUpdateSet :: UpdateSet -> UpdateSet -> UpdateSet
+concatUpdateSet (UpdateSet cres1 arcs1 chos1) (UpdateSet cres2 arcs2 chos2) =
+  UpdateSet (cres1 ++ cres2) (arcs1 ++ arcs2) (chos1 ++ chos2)
+
+-- | The environment for the DAML-LF verifier
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:71]: Is there some deeper meaning behind calling this `Delta` apart from the fact that this is commonly used in type theory? If not, maybe just call it `Environment` or some variation thereof?

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+-- The issue here is that our values are currently stored in Delta instead
+-- of in the world like in the type checker.
+-- This means that we can't throw the error we want to throw here.
+-- 2 options:
+--   + either define our own errors, as they don't correspond 100% to the type
+--   checking errors anyway.
+--   + or create our own world environment to store values. This also makes
+--   sense as these values work similar to how they work in the type checker,
+--   except that we need to store a partially evaluated definition as well.
+-- Both approaches make sense, but both imply a lot of code duplication, so they
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:125]: I wouldn’t worry too much about duplicating code for now.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            -> m Delta
+genPackages inp = ask >>= \del0 -> foldM genPackage' del0 inp
+  where
+    -- TODO: These prime definitions are bit silly. Merge with the regular
+    -- functions.
+    -- TODO: This >>= return . concatDelta returns a lot. Figure out a way to
+    -- abstract over it. Same for this returning foldM.
+    genPackage' :: MonadDelta m => Delta
+                -> (PackageId, (Package, Maybe PackageName)) -> m Delta
+    genPackage' deli (idi, (paci, _)) =
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:76]: Why do all variable names end with `i`? :slightly_smiling_face:

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+      -- TODO: This is getting quite unreadable.
+      foldM (\deli' modi -> genModule' deli' (PRImport idi, modi))
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:78]: I’m probably missing something but it looks to me like this only works if modules are sorted topologically. Otherwise, you might not be able to resolve a cross-module reference since you haven’t yet included the delta for that module?

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    genTemplate' deli temi = introDelta deli (genTemplate temi)
+                          >>= return . (concatDelta deli)
+
+genValue :: MonadDelta m => PackageRef -> ModuleName -> DefValue -> m Delta
+genValue pac mod val = do
+  expOut <- genExpr (dvalBody val)
+  let qname = Qualified pac mod (fst $ dvalBinder val)
+  return $ set devals (singleton qname (expOut ^. goExp, expOut ^. goUpd)) emptyDelta
+
+-- TODO: Handle annotated choices, by returning a set of annotations.
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:106]: What is an annotated choice? Presumably that’s some new annotation that you want to add to express constraints that should be verified?

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+genForCreate tem arg = do
+  argOut <- genExpr arg
+  case argOut ^. goExp of
+    argExpr@(ERecCon _ fs) -> return (GenOutput (EUpdate (UCreate tem argExpr))
+                                      -- TODO: We could potentially filter here
+                                      -- to only store the interesting fields?
+                                      (set usCre [UpdCreate tem fs] emptyUpdateSet)
+                                      (argOut ^. goDel))
+    _                      -> throwError EEnumTypeWithParams
+    -- TODO: This is a random error, as we do not have access to the expected
```
Comment (cocreature) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:176]: I’m perfectly happy with just calling `error` for now for anything that isn’t well-typed LF.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+-- environment extension.
+introDelta :: MonadDelta m => Delta -> m a -> m a
+introDelta delta = local (concatDelta delta)
+
+extVarDelta :: MonadDelta m => ExprVarName -> m a -> m a
+extVarDelta x = local (over devars ((:) x))
+
+lookupDExprVar :: MonadDelta m => ExprVarName -> m ()
+lookupDExprVar x = ask >>= \ del -> unless (elem x $ del ^. devars)
+                                          $ throwError $ EUnknownExprVar x
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:110]: I don't think lenses buy us anything here. I also think `a >>= \x -> ...` is a sign to use `do` notation.
```suggestion
lookupDExprVar x = do
  del <- ask
  unless (x `elem` _devars del) $ throwError $ EUnknownExprVar x
```

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+-- | Extend a generator output with the updates and the context extensions of a
+-- second generator output. Note that the end result will contain the first
+-- expression.
+combineGO :: GenOutput -> GenOutput
+          -- ^ The two generator outputs to combine.
+          -> GenOutput
+combineGO genOut1 genOut2
+  = extendGOUpds (genOut2 ^. goUpd)
+  $ extendGODelta (genOut2 ^. goDel)
+    genOut1
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:43]: Isn't that just plain function application?
```suggestion
  = extendGOUpds (_goUpd genOut2)
  $ extendGODelta (_goDel genOut2)
    genOut1
```

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+    -- functions.
+    -- TODO: This >>= return . concatDelta returns a lot. Figure out a way to
+    -- abstract over it. Same for this returning foldM.
+    genPackage' :: MonadDelta m => Delta
+                -> (PackageId, (Package, Maybe PackageName)) -> m Delta
+    genPackage' deli (idi, (paci, _)) =
+      -- TODO: This is getting quite unreadable.
+      foldM (\deli' modi -> genModule' deli' (PRImport idi, modi))
+            deli (NM.toList $ packageModules paci)
+      >>= return . (concatDelta deli)
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:80]: `act >>= return . f` is the same as `fmap f act` or `f <$> act`, depending on what you prefer.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+genTemplate :: MonadDelta m => Template -> m Delta
+genTemplate = undefined -- TODO
+
+genExpr :: MonadDelta m => Expr -> m GenOutput
+genExpr = \case
+  ETmApp fun arg  -> genForTmApp fun arg
+  ETyApp expr typ -> genForTyApp expr typ
+  EVar name       -> genForVar name
+  EVal w          -> genForVal w
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:127]: Please don't align the arrows in case expressions. This will lead to a lot of "diff noise" down the line.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+                          >>= return . (concatDelta deli)
+    genTemplate' :: MonadDelta m => Delta -> Template -> m Delta
+    genTemplate' deli temi = introDelta deli (genTemplate temi)
+                          >>= return . (concatDelta deli)
+
+genValue :: MonadDelta m => PackageRef -> ModuleName -> DefValue -> m Delta
+genValue pac mod val = do
+  expOut <- genExpr (dvalBody val)
+  let qname = Qualified pac mod (fst $ dvalBinder val)
+  return $ set devals (singleton qname (expOut ^. goExp, expOut ^. goUpd)) emptyDelta
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:104]:
```suggestion
  return $ emptyDelta{_devals = singleton qname (_goExp expOut, _goUpd expOut)}
```

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+              -- ^ The current generator output.
+              -> GenOutput
+extendGODelta delta = over goDel (concatDelta delta)
+
+-- | Builds up the Delta environment by recursively extending the environment
+-- for each computation, and combining the output environments.
+buildDelta :: MonadDelta m => Delta -> (a -> m Delta) -> [a] -> m Delta
+buildDelta del op args = foldM step del args
+  where
+    step d x = concatDelta d <$> setDelta d (op x)
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:71]: My _suspicion_, fuelled by your comment on `setDelta`, is that you want something like this:
```haskell
introDeltas :: MonadDelta m => (a -> m Delta) -> [a] -> m b -> m b
introDeltas _ [] act = act
introDeltas f (x:xs) act = do
  delta <- f x
  introDelta delta (introDeltas f xs act)
```
and then have functions like
```haskell
introModule :: MonadDelta m => PackageRef -> Module -> m b -> m b
```
Is that correct?

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs (hunk 10 above)
Comment (hurryabit) [reply compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:166]: I usually use something ala `error "IMPOSSIBLE: Ill-typed DAML-LF"` in situations like this one.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+-- | The List of updates being performed
+data UpdateSet = UpdateSet
+  { _usCre :: ![UpdCreate]
+    -- ^ The list of create updates.
+  , _usArc :: ![UpdArchive]
+    -- ^ The list of archive updates.
+  , _usCho :: ![UpdChoice]
+    -- ^ The list of choice updates.
+  }
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:60]: Don't we need to account for fetches as well?

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs (hunk 10 above)
Comment (cocreature) [reply compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs:166]: The impossible happened :open_mouth:

**Comment by cocreature** [comment]
@Gertjan423  CI is currently failing due to the format check. You can easily fix that by running `./fmt.sh` locally or if you want to have it be a bit faster `./fmt.sh --diff`. The latter is fast enough that some people even set it up for a precommit or at least prepush hook.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify.hs [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+fieldName :: FieldName
+fieldName = FieldName "amount"
+
+main :: IO ()
+main = do
+  Options{..} <- execParser optionsParserInfo
+  solver <- locateRunfiles "z3_nix/bin/z3"
+  pkgs <- readPackages optInputDars
+  putStrLn "Start value phase" >> case runEnv (genPackages ValuePhase pkgs) emptyEnv of
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify.hs:31]:
```suggestion
  putStrLn "Start value phase"
  case runEnv (genPackages ValuePhase pkgs) emptyEnv of
```

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+  UpdateSet (cres1 ++ cres2) (arcs1 ++ arcs2) (chos1 ++ chos2) (vals1 ++ vals2)
+
+-- | Refresh a given expression variable by producing a fresh renamed variable.
+-- TODO: when a renamed var gets renamed again, it might overlap again.
+-- We should have an additional field in VarName to denote it's number.
+genRenamedVar :: MonadEnv m
+  => ExprVarName
+  -- ^ The variable to be renamed.
+  -> m ExprVarName
+genRenamedVar (ExprVarName x) = ExprVarName . T.append x . T.pack <$> fresh
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:106]: This does not necessarily rename to a fresh name. For instance, `x` could be renamed to `x1` which is already in use. Is this intended?

File: compiler/daml-lf-verify/BUILD.bazel [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+        "//compiler/daml-lf-tools",
+        "//libs-haskell/bazel-runfiles",
+        "//libs-haskell/da-hs-base",
+    ],
+)
+
+da_haskell_test(
+    name = "verify-tests",
+    srcs = glob(["tests/**/*.hs"]),
+    data = glob(["data/*.dar"]),
```
Comment (hurryabit) [inline compiler/daml-lf-verify/BUILD.bazel:94]: We should not check in the DAR files directly but rather the DAML source. You can use the `daml_compile` rule to produce the DARs then. You can find an example at https://github.com/digital-asset/daml/blob/9b476416b8b16ee936e10cc97f11dd1461f8b173/language-support/ts/codegen/tests/BUILD.bazel#L12.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify.hs [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+  pkgs <- readPackages [dar]
+  solver <- getSolver
+  when verbose $ putStrLn "Start value gathering"
+  case runEnv (genPackages pkgs) (emptyEnv :: Env 'ValueGathering) of
+    Left err-> do
+      putStrLn "Value phase finished with error: "
+      print err
+      return Unknown
+    Right env1 -> do
+      when verbose $ putStrLn "Start value solving"
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify.hs:60]: Maybe it's easier to pass a `debug :: String -> IO ()` function rather than `verbose`. In normal mode you pass `\_ -> pure ()` and in debug mode `putStrLn`.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Read.hs [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+data Options = Options
+    { optInputDar :: FilePath
+    , optChoiceTmpl :: String
+    , optChoiceName :: String
+    , optFieldTmpl :: String
+    , optFieldName :: String
+    }
+
+optionsParser :: Parser Options
+optionsParser = Options
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Read.hs:35]: Having so many positional arguments will probably become an issue in the future because it's easy to mix them up. But let's fix that in a follow-up PR.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 23]
```diff
@@ -0,0 +1,812 @@
...
+data UpdChoice = UpdChoice
+  { _choTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _choName  :: !ChoiceName
+    -- ^ The name of the choice.
+  }
+  deriving (Eq, Generic, Hashable, Show)
+
+-- | The collection of updates being performed.
+data UpdateSet (ph :: Phase) where
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:148]: My impression is that you're using GADTs here to link the `UpdateSet` types from the three different phases with their `Env` counterparts. This problem might be better solved with associated data types similar to
```haskell
class IsPhase (ph :: Phase) where
  data UpdateSet ph
  data Env ph
```
Let's try that out in a follow-up PR.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 24]
```diff
@@ -0,0 +1,812 @@
...
+  cid <- expr2cid exp
+  cids <- getEnv >>= \case
+    EnvVG{..} -> return _envvgcids
+    EnvCG{..} -> return _envcgcids
+    EnvS{..} -> return _envscids
+  case HM.lookup cid cids of
+    Nothing -> throwError $ UnknownCid cid
+    Just var -> return var
+
+-- TODO: There seems to be something wrong with the PackageRef in Qualified.
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:543]: Are you aware that there is `PRSelf` for references into the same package? You might want to rewrite all these references into `PRImport <own-package-id>` during package loading.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 25]
```diff
@@ -0,0 +1,812 @@
...
+-- | Helper function to lookup a value definition in a HashMap.
+lookupValInHMap :: HM.HashMap (Qualified ExprValName) (Expr, UpdateSet ph)
+  -- ^ The HashMap in which to look.
+  -> Qualified ExprValName
+  -- ^ The value name to lookup.
+  -> Maybe (Expr, UpdateSet ph)
+lookupValInHMap hmap val = listToMaybe $ HM.elems
+  $ HM.filterWithKey (\name _ -> qualObject name == qualObject val && qualModule name == qualModule val) hmap
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:551]: Don't we have `(==)` in `Qualified`? But maybe it's relating to your the package reference problem mentioned above.

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs [hunk 26]
```diff
@@ -0,0 +1,812 @@
...
+      let (upd_ia, hmap_ia) = solveReference lookup getRefs extUpds introCond (ref0:vis) hmap_i0 ref_ia
+          (upd_ib, hmap_ib) = maybe (Nothing, hmap_ia)
+            (first Just . solveReference lookup getRefs extUpds introCond (ref0:vis) hmap_ia) ref_ib
+          upd_i1 = extUpds upd_i0 $ introCond $ Conditional cond upd_ia upd_ib
+      in (upd_i1, hmap_ib)
+
+-- TODO: This should work recursively
+-- | Lookup the field names and corresponding types, for a given record type
+-- constructor name.
+recTypConFields :: MonadEnv m ph
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs:714]: `rec` is not the best prefix when both "record" and "recursion" are common words in your problem domain. 😄

File: compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Subst.hs [hunk 27]
```diff
@@ -0,0 +1,157 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- | Term substitions for DAML LF static verification
+module DA.Daml.LF.Verify.Subst
```
Comment (hurryabit) [inline compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Subst.hs:5]: We're about to land another implementation of substitution for DAML-LF in https://github.com/digital-asset/daml/pull/5759. We should switch the code here over to use @sofiafaro-da's implementation once it has landed.

File: compiler/daml-lf-verify/tests/DA/Daml/LF/Verify/Tests.hs [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+        Success result
+  , testCase "Fail A" $ do
+      condDar <- locateRunfiles (mainWorkspace </> "compiler/daml-lf-verify/data/conditionals.dar")
+      let tmpl = TypeConName ["Iou"]
+          choice = ChoiceName "FailA"
+          field = FieldName "content"
+      verify condDar False tmpl choice tmpl field >>= \case
+        Success -> assertFailure "Verification wrongfully passed for FailA - content"
+        Unknown -> assertFailure "Verification inconclusive for FailA - content"
+        Fail _ -> return ()
```
Comment (cocreature) [inline compiler/daml-lf-verify/tests/DA/Daml/LF/Verify/Tests.hs:87]: Ar the results not deterministic or why are you not just using `assertEqual` on the full `Fail` expression?

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs [hunk 29]
```diff
@@ -3,7 +3,7 @@
 
 -- | DAML-LF Numeric literals, with scale attached.
 module DA.Daml.LF.Ast.Numeric
-    ( Numeric
+    ( Numeric (..)
```
Comment (hurryabit) [inline compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs:6]: I don't think this is a good idea. Why did you change this?

File: compiler/daml-lf-verify/daml/conditionals/.gitignore [hunk 30]
```diff
@@ -0,0 +1 @@
+/.daml
```
Comment (hurryabit) [inline compiler/daml-lf-verify/daml/conditionals/.gitignore:1]: This can go.

File: compiler/daml-lf-verify/daml/conditionals/daml.yaml [hunk 31]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (hurryabit) [inline compiler/daml-lf-verify/daml/conditionals/daml.yaml:1]: I'm not 100% sure we need the `daml.yaml` files for the test examples. Can you please check if the tests still work without them and remove them if possible. Same further below.

File: compiler/daml-lf-verify/daml/quickstart/.gitignore [hunk 32]
```diff
@@ -0,0 +1 @@
+/.daml
```
Comment (hurryabit) [inline compiler/daml-lf-verify/daml/quickstart/.gitignore:1]: This can go.

File: compiler/daml-lf-verify/daml/quickstart/src/main/java/com/digitalasset/quickstart/iou/IouMain.java [hunk 33]
```diff
@@ -0,0 +1,143 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (hurryabit) [inline compiler/daml-lf-verify/daml/quickstart/src/main/java/com/digitalasset/quickstart/iou/IouMain.java:1]: This can go.

File: compiler/daml-lf-verify/daml/quickstart/src/main/resources/logback.xml [hunk 34]
```diff
@@ -0,0 +1,17 @@
+<configuration>
```
Comment (hurryabit) [inline compiler/daml-lf-verify/daml/quickstart/src/main/resources/logback.xml:1]: This can go.

File: compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs (hunk 29 above)
Comment (hurryabit) [reply compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs:6]: Actually, this is fine. I cannot read and thought we were changing a DAML file here, which we obviously aren't. So it's all good. Apologies for the confusion I caused.

**Comment by hurryabit** [comment]
@Gertjan423 You need to run `./fmt.sh` in the root of the repo again.

File: compiler/daml-lf-verify/BUILD.bazel [hunk 35]
```diff
@@ -46,7 +46,7 @@ da_haskell_binary(
         "//libs-haskell/bazel-runfiles",
         "//libs-haskell/da-hs-base",
     ],
-)
+) if not is_windows else None
```
Comment (cocreature) [inline compiler/daml-lf-verify/BUILD.bazel:50]: Why do we need to disable everything on Windows? I get that we need to disable the tests until we get Z3 from somewhere but building things seems like it should be fine?

Comment (Gertjan423) [reply compiler/daml-lf-verify/BUILD.bazel]: That's a fair point. Fixed, now only the tests are disabled :)
